modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java

```diff
@@ -1382,4 +1394,94 @@ protected boolean isPostgisDialect() {
+        preQuery.setQueryJoins(((JoiningQuery) query).getQueryJoins());
+
+        // Build the feature type returned by this query. Also build an eventual extra feature type
+        // containing the attributes we might need in order to evaluate the post filter
+        SimpleFeatureType querySchema;
+        if (query.getPropertyNames() == Query.ALL_NAMES) {
+            querySchema = getSchema();
+        } else {
+            querySchema = SimpleFeatureTypeBuilder.retype(getSchema(), query.getPropertyNames());
+        }
+
+        // grab connection
+        JDBCDataStore store = getDataStore();
+        Connection cx = store.getConnection(getState().getTransaction());
+        Statement st = null;
+        ResultSet rs = null;
+        try {
+
+            SQLDialect dialect = store.getSQLDialect();
+
+            if (dialect instanceof PreparedStatementSQLDialect) {
+                AtomicReference<PreparedFilterToSQL> toSQLref = new AtomicReference<>();
+                String sql = createCountQuery(dialect, querySchema, jQuery, toSQLref);
+                st =
+                        cx.prepareStatement(
+                                sql, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY);
+                st.setFetchSize(getDataStore().fetchSize);
+                rs = ((PreparedStatement) st).executeQuery();
+            } else {
+                String sql = createCountQuery(dialect, querySchema, jQuery, null);
+                st = cx.createStatement();
+                rs = st.executeQuery(sql);
+            }
+            int result = 0;
+            if (rs.next()) {
+                Object value = rs.getObject(1);
+                result = Converters.convert(value, Integer.class);
+            }
+            int maxFeatures = query.getMaxFeatures();
+            if (maxFeatures > 0 && result > maxFeatures) result = query.getMaxFeatures();
+            return result;
+        } catch (Exception e) {
+            // safely rethrow
+            throw (IOException) new IOException().initCause(e);
+        } finally {
+            store.closeSafe(rs);
+            store.closeSafe(st);
+            store.closeSafe(cx);
+        }
+    }
+
+    private String createCountQuery(
+            SQLDialect dialect,
+            SimpleFeatureType querySchema,
+            JoiningQuery query,
+            AtomicReference<PreparedFilterToSQL> toSQLRef)
+            throws IOException, SQLException, FilterToSQLException {
+        String idColumnName = getIdColumnName(getSchema());
+        StringBuffer countSQL = new StringBuffer("SELECT COUNT(").append("DISTINCT ");
+        dialect.encodeColumnName(COUNT_TABLE_ALIAS, idColumnName, countSQL);
+        countSQL.append(")").append(" FROM (");
+        String sql = selectSQL(querySchema, query, toSQLRef, true);
+        countSQL.append(sql).append(") ");
+        dialect.encodeTableName(COUNT_TABLE_ALIAS, countSQL);
+        LOGGER.fine(countSQL.toString());
```

Comment:
countSQL.toString() is evaluated twice and always even if LOGGER level is not DEBUG.  Better save the built string in a variable and use it on both cases.

Reply:
fixed